docs/madr/decisions/042-tcp-grpc-probe.md

```diff
@@ -0,0 +1,246 @@
+      httpGet:
+        path: /healthz
+        port: 8080
+    livenessProbe:
+      grpc:
+        port: 2379
+        service: liveness
+    readinessProbe:
+      tcpSocket:
+        port: 1433
+```
+
+In which:
+- `http` is the port to be used for all the HTTP probes, replacing the existing annotation `kuma.io/virtual-probes-port`
+- `grpc-*` is the ports to be used for each of gRPC probes
+- `tcp-*` is the ports to be used for each of TCP probes
+
+If a probe is not specified in the `kuma.io/virtual-probes-port-mapping`, we'll generate the port automatically by incrementing from the default port (`9000`), skipping the known and taken ports.
+
+#### Virtual Probes Listeners
+
+We generate a TCP proxy listener for each of gRPC and TCP probes, and forward the probe traffic transparently to the application. We also keep the existing HTTP Virtual Probe listener for HTTP probes.
+
+This keep the compatibility with the existing HTTP based Virtual Probe and data plane data model.
+
+#### Data Plane Data Model 
+
+Introduce two new fields onto the `Dataplane` resource type to store the Virtual Probe ports:
+
+```yaml
+  probes:
+    endpoints:
+    - inboundPath: /metrics
+      inboundPort: 8081
+      path: /8081/metrics
+    port: 19000
+  tcpProbes:
+  - port: 19001
+    applicationPort: 1433
+  grpcProbes:
+  - port: 19002
+    applicationPort: 2379
+    service: abcd
+```
+
+#### Positive Consequences
+
+- Ease of implementation
+
+- Keeping the compatibility with the existing HTTP based Virtual Probe and data plane data model
+
+#### Negative Consequences
+
+- The user experience of specifying virtual probes ports is more complex
+
+- More virtual listeners on sidecar
+
+## Other options
+
+### Merging all probes into one listener
+
+Introducing an intermediate layer to translate HTTP probes into TCP and gRPC probes (the translator), transform user specified probes (the user probe) into HTTP probes on the pod, and:
+
+1. If the user probe is HTTP based, requests are forwarded/redirected to application directly
+
+2. If the user probe is TCP or gRPC based, we'll still receive HTTP requests for these probes, since they are transformed to HTTP probes. We forward these requests to the translator and the translator is responsible for detecting the actual probe status from the application.
+
+The whole workflow can be shown in the following diagram:
+
+```
+  gRPC Probes                                               HTTP to gRPC translator --->  Application
+              ↘                                           ↗
+HTTP Probes --> HTTP Probes --->  Virtual Probe Listener ->  HTTP Probes  --->  Application
+              ↗                                           ↘
+   TCP Probes                                               HTTP to TCP translator  --->  Application
+```
+
+#### Positive Consequences
+
+- Don't need to introduce new ports for TCP and gRPC probes
+
+- Better user experience: don't need to introduce new annotations
```

Comment:
It's worth mentioning that with the existing annotation folks only need to play with it if they use port `19000` for their app

Reply:
Added onto line 48